project/src/main/puzzle/level/rule-parser.gd

```diff
@@ -1,51 +1,427 @@
 """
-func float_value() -> float:
-	return float(_parsed_rules[_prev_key])
+func add_bool(name: String, false_string: String = "") -> OngoingPropertyDefinition:
+	return add(BoolPropertyParser.new(_target, name, false_string))
 
 
 """
-Returns the string value for the key previously passed into the 'has' function.
+Adds an enum property parser.
+
+Parameters:
+	'name': The name of the property and json key to parse.
+	
+	'enum_dict': The property's enum type, such as 'PuzzleTileMap.TileSetType'
 """
-func string_value() -> String:
-	return str(_parsed_rules[_prev_key])
+func add_enum(name: String, enum_dict: Dictionary) -> OngoingPropertyDefinition:
+	return add(EnumPropertyParser.new(_target, name, enum_dict))
+
 
+"""
+Adds a float property parser.
+
+Parameters:
+	'name': The name of the property and json key to parse.
+"""
+func add_float(name: String) -> OngoingPropertyDefinition:
+	return add(FloatPropertyParser.new(_target, name))
+
+
+"""
+Adds an int property parser.
+
+Parameters:
+	'name': The name of the property and json key to parse.
+"""
+func add_int(name: String) -> OngoingPropertyDefinition:
+	return add(IntPropertyParser.new(_target, name))
+
+
+"""
+Adds a string property parser.
+
+Parameters:
+	'name': The name of the property and json key to parse.
+"""
+func add_string(name: String) -> OngoingPropertyDefinition:
+	return add(StringPropertyParser.new(_target, name))
+
+
+"""
+Adds a custom property parser.
+
+Parameters:
+	'property_parser': Defines behavior for parsing and formatting json strings.
+"""
+func add(property_parser: PropertyParser) -> OngoingPropertyDefinition:
+	_property_parsers.append(property_parser)
+	for rule_key in property_parser.keys:
+		_property_parsers_by_key[rule_key] = property_parser
+	var ongoing_rule_definition := OngoingPropertyDefinition.new(property_parser)
+	return ongoing_rule_definition
+
+
+"""
+Updates the rule's properties based on the specified json strings.
+
+Parameters:
+	'json': An array of json strings to parse, such as 'tile_set veggie'.
+"""
+func from_json_array(json: Array) -> void:
+	for json_string in json:
+		var split: Array = json_string.split(" ")
+		if not _property_parsers_by_key.has(split[0]):
+			push_warning("Unrecognized: %s" % [split[0]])
+			continue
+		var property_parser: PropertyParser = _property_parsers_by_key[split[0]]
+		property_parser.from_json_string(json_string)
+
+
+"""
+Returns a minimal list of json strings corresponding to the rule's current values.
+
+Properties which are set to their default value are omitted from the resulting list. If all properties are assigned to
+their default value, the list will be empty.
+"""
+func to_json_array() -> Array:
+	var result := []
+	for property_parser in _property_parsers:
+		if not property_parser.is_default():
+			result.append_array(property_parser.to_json_strings())
+	return result
+
+
+"""
+Returns 'true' if all of the rule's properties are currently set to their default values.
```

Comment:
It might be good to hint at why we do this.

Reply:
Espanded comment, `Rules whose properties are set to their default values are omitted from our json output.`